feat: documentation of codemods (#19055)
carbon-design-system/carbon: packages/react/src/components/FeatureFlags/overview.mdx

| -------------------------------------------------- | ------------------------------------------------------------------------------------ | ------- | --------------- | --------- | ----------------- |
| `enable-experimental-tile-contrast`                | Enable the improved styling for tiles that provides better contrast                  | `false` |                 | ✅        | No - Style only   |
| `enable-experimental-focus-wrap-without-sentinels` | Enable the new focus wrap behavior that doesn't use sentinel nodes                   | `false` | ✅              |           | No - Config only  |
| `enable-dialog-element`                            | Enable components to utilize the native dialog element                               | `false` | ✅              | ✅        | No - Config only  |
| `enable-treeview-controllable`                     | Enable the new TreeView controllable API                                             | `false` | ✅              |           | No - Config only  |

> | `enable-v12-tile-default-icons`                    | Enable default icons for Tile components                                             | `false` | ✅              |           | [Yes](/docs/getting-started-v12--overview#enable-v12-tile-default-icons) |

Reviewer: Adding the `codemod available` column is a great idea. These can link to the section heading for this codemod over in the `@carbon/upgrade` readme. Could we use the "name" of the codemod (like what you put into the command) as the link instead of "Yes"?

| `enable-v12-overflowmenu`                          | Enable the use of the v12 OverflowMenu leveraging the Menu subcomponents             | `false` | ✅              |           | [Yes](/docs/getting-started-v12--overview#enable-v12-overflowmenu) |
| `enable-v12-tile-radio-icons`                      | Enable rendering of default icons in the tile components                             | `false` | ✅              | ✅        | [Yes](/docs/getting-started-v12--overview#enable-v12-tile-radio-icons) |
| `enable-v12-structured-list-visible-icons`         | Enable icon components within StructuredList to always be visible                    | `false` |                 | ✅        | [Yes](/docs/getting-started-v12--overview#enable-v12-structured-list-visible-icons) |
| `enable-v12-dynamic-floating-styles`               | Enable dynamic setting of floating styles for components like Popover, Tooltip, etc. | `false` | ✅              |           | No - App level    |

> Some feature flags don't have codemods because:

Reviewer: What if we added a section near the bottom of this file to house this content and describe what "codemod available" means. Some of it could be lifted from the content in the v12 page:

---

> ## Using codemods for migration
> 
> Codemods are code modification scripts that automate the necessary changes when migrating to new components and APIs. They reduce the manual effort required when adopting new features. Carbon provides codemods to help you enable feature flags and migrate to the new APIs flags provide. 
> 
> Some feature flags don't have codemods because:
> - ...
> 
> For more information on codemods, see the [`@carbon/upgrade` documentation](https://github.com/carbon-design-system/carbon/tree/main/packages/upgrade#carbonupgrade)

> - **Style only**: The flag only affects CSS styling with no component API changes

Reviewer: ```suggestion
- **Style only**: The flag only affects Sass styles with no changes to the javascript component API.
```

> - **Config only**: The flag is a configuration option that users enable at selected places

Reviewer: Which flags fall into this category? I think this could be more descriptive but am not sure what flags you're thinking about and so I'm unsure what to recommend.

> - **App level**: The flag is typically applied at the application level in just a few places

Reviewer: ```suggestion
- **App level**: The flag is intended to be applied near or at the application root
```

@include feature-flags.enable('enable-experimental-tile-contrast');
```
## Using Codemods for Migration

Codemods are code modification scripts that automate the necessary changes when migrating to new components and APIs. They reduce the manual effort required when adopting new features. Carbon provides codemods to help you enable feature flags and migrate to the new APIs these flags provide.

Some feature flags don't have codemods because:

> - **Style only**: Flags like `enable-experimental-tile-contrast` only affect Sass styles with no changes to the javascript component API, so no code transformation is needed.

Reviewer: ```suggestion
- **Style only**: Flags like `enable-experimental-tile-contrast` only affect Sass styles with no changes to the javascript component API.
```

- **Config only**: Flags like `enable-experimental-focus-wrap-without-sentinels`, `enable-dialog-element`, and `enable-treeview-controllable` are configuration options that are enabled in just a few specific places in your application.

> - **App level**: Flags like `enable-v12-dynamic-floating-styles` are designed to be applied at the application level to enable specific behaviors (like floating styles without full autoAlign behavior) and don't require widespread component changes.

Reviewer: Like we talked about, I think we could merge these? I don't know what title to call it, but the description could be something like

> These flags don't have codemods because their usage can vary widely. Some will opt to turn them on for a single component, some will turn them on for a sub-section(s) of their app/react tree, and some will enable it globally at their root.
